apm-agent-core/src/main/java/co/elastic/apm/agent/util/IOUtilsProviderImpl.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.util;
+
+import co.elastic.apm.agent.objectpool.ObjectHandle;
+import co.elastic.apm.agent.objectpool.ObjectPool;
+import co.elastic.apm.agent.objectpool.ObjectPoolFactory;
+import co.elastic.apm.agent.sdk.internal.util.IOUtils;
+import co.elastic.apm.agent.tracer.pooling.Allocator;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CoderResult;
+import java.nio.charset.StandardCharsets;
+
+public class IOUtilsProviderImpl implements IOUtils.IOUtilsProvider {
+
+    protected static final int BYTE_BUFFER_CAPACITY = 2048;
+
+    private static class DecoderWithBuffer {
+        final ByteBuffer byteBuffer = java.nio.ByteBuffer.allocate(BYTE_BUFFER_CAPACITY);
+        final CharsetDecoder decoder = StandardCharsets.UTF_8.newDecoder();
+    }
+
+    private final ObjectPool<ObjectHandle<DecoderWithBuffer>> pool;
+
+    public IOUtilsProviderImpl() {
+        this.pool = new ObjectPoolFactory().createHandlePool(new Allocator<DecoderWithBuffer>() {
+            @Override
+            public DecoderWithBuffer createInstance() {
+                return new DecoderWithBuffer();
+            }
+        });
+    }
+
+    @Override
+    public boolean readUtf8Stream(final InputStream is, final CharBuffer charBuffer) throws IOException {
+        // to be compatible with Java 8, we have to cast to buffer because of different return types
+        try (ObjectHandle<DecoderWithBuffer> pooled = pool.createInstance()) {
```

Comment:
[For Reviewer] Unfortunately the diff does not recognize it, but the actual implementations have not been touched. The only change was to replace the `ThreadLocal` usage with the resource pool.